### PR TITLE
fix(#2355): cap tedge-watchdog interval to 1s minimum

### PR DESCRIPTION
**Pull request by Bravo555**
## Proposed changes

Cap tedge-watchdog interval in which we expect to receive health check response to a minimum of 1s.

I also had questions of "why is the interval one-fourth of what we read from `WatchdogSec`?", so decided to make it a constant and document it a little more clearly. Found that this ratio was changed [here](https://github.com/thin-edge/thin-edge.io/pull/1060) and [here](https://github.com/thin-edge/thin-edge.io/pull/1170), but the PRs don't mention why these specific values were chosen (and links to issues are missing).

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

- #2355 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

### CODE REVIEW

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 1]
```diff
@@ -61,7 +68,10 @@ async fn start_watchdog_for_self() -> Result<(), WatchdogError> {
                     let _ = notify_systemd(process::id(), "WATCHDOG=1").map_err(|e| {
                         eprintln!("Notifying systemd failed with {}", e);
                     });
-                    tokio::time::sleep(tokio::time::Duration::from_secs(interval / 4)).await;
+                    let delay = tokio::time::Duration::from_secs(
+                        (interval / NOTIFY_SEND_FREQ_RATIO).max(1),
```
Comment (didier-wenzek) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:72]: Don't you mean:

```suggestion
                        (interval / NOTIFY_SEND_FREQ_RATIO).min(1),
```

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 2]
```diff
@@ -129,12 +138,14 @@ async fn start_watchdog_for_tedge_services(tedge_config_dir: PathBuf) {
 
                 let tedge_config_location = tedge_config_location.clone();
                 watchdog_tasks.push(tokio::spawn(async move {
+                    //
+                    let interval = Duration::from_secs((interval / NOTIFY_SEND_FREQ_RATIO).max(1));
```
Comment (didier-wenzek) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:142]:
```suggestion
                    let interval = Duration::from_secs((interval / NOTIFY_SEND_FREQ_RATIO).min(1));
```

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge_watchdog/src/systemd_watchdog.rs:72]: Oops, you are correct !

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 3]
```diff
@@ -33,6 +33,13 @@ use tracing::warn;
 
 const SERVICE_NAME: &str = "tedge-watchdog";
 
+/// How many times more often do we send notify to systemd watchdog, than is necessary from the
+/// `WatchdogSec` value.
```
Comment (albinsuresh) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:37]: It couldn't be `WatchdogSec` value itself, as there's the risk of missing notifications due to the additional delay between the health check request and response. For e.g, assume a `WatchdogSec` value of `30s` and the systemd watchdog timer starts at `T`. If our tedge_watchdog wakes up from its sleep at `T+29` and sends the health check request, the response might be slightly delayed and received at `T+31` leading to the watchdog timer expiring leading to a restart.

That's why we started with half of `WatchdogSec` which gives sufficient buffer for the responses to arrive. Theoretically that should have worked, but in practice, even that interval seemed insufficient and we still encountered occasional notification misses, which lead to the quarter value.

BTW, even smaller fractions won't work if the `WatchdogSec` value itself is too small, like a couple of seconds, as the resultant computed delay of 1 sec might also not work, if the request-response delay is higher.

**Review by albinsuresh — APPROVED** [review]

**Pull request merged**